content/courses/advanced-ada/chapters/numerics.rst

```diff
@@ -1639,12 +1639,747 @@ precision, 0.2774 is rounded to 0.277, and 0.2777 is rounded to 0.278.
+
+    pragma Ada_2022;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Ada.Numerics.Big_Numbers.Big_Integers;
+    use  Ada.Numerics.Big_Numbers.Big_Integers;
+
+    with Ada.Numerics.Big_Numbers.Big_Reals;
+    use  Ada.Numerics.Big_Numbers.Big_Reals;
+
+    procedure Show_Simple_Big_Numbers is
+       BI : Big_Integer;
+       BR : Big_Real;
+    begin
+       BI := 12345678901234567890;
+       BR := 2.0 ** 1234;
+
+       Put_Line ("BI: " & BI'Image);
+       Put_Line ("BR: " & BR'Image);
+
+       BI := @ + 1;
+       BR := @ + 1.0;
+
+       Put_Line ("BI: " & BI'Image);
+       Put_Line ("BR: " & BR'Image);
+    end Show_Simple_Big_Numbers;
+
+In this example, we're declaring the big integer :ada:`BI` and the big real
+:ada:`BR`, and we're incrementing them with one.
+
+Naturally, we're not limited to using the :ada:`+` operator (such as in this
+example). We can use any operators on big numbers in the same as we do with
+integers and floating-point variables. In fact, the common unary operators
+(:ada:`+`, :ada:`-`, :ada:`abs`) and binary operators (:ada:`+`, :ada:`-`,
+:ada:`*`, :ada:`/`, :ada:`**`, :ada:`Min` and :ada:`Max`) are available to us.
+For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Numerics.Simple_Big_Numbers_Operators
+
+    pragma Ada_2022;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Ada.Numerics.Big_Numbers.Big_Integers;
+    use  Ada.Numerics.Big_Numbers.Big_Integers;
+
+    procedure Show_Simple_Big_Numbers_Operators is
+       BI : Big_Integer;
+    begin
+       BI := 12345678901234567890;
+
+       Put_Line ("BI: " & BI'Image);
+
+       BI := -@ + BI / 2;
+       BI :=  @ - BI * 2;
+
+       Put_Line ("BI: " & BI'Image);
+    end Show_Simple_Big_Numbers_Operators;
+
+In this example, we're applying the four basic operators (:ada:`+`, :ada:`-`,
+:ada:`*`, :ada:`/`) on big integers.
+
+
+Factorial
+~~~~~~~~~
+
+A typical example is the :wikipedia:`factorial <Factorial>`: a sequence of the
+factorial of consecutive small numbers can quickly lead to big numbers. Let's
+take this implementation as an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Numerics.Factorial_Integer
+    :class: ada-run-expect-failure
+
+    function Factorial (N : Integer)
+                        return Long_Long_Integer;
+
+    function Factorial (N : Integer)
+                        return Long_Long_Integer is
+       Fact : Long_Long_Integer := 1;
+    begin
+       for I in 2 .. N loop
+          Fact := Fact * Long_Long_Integer (I);
+       end loop;
+
+       return Fact;
+    end Factorial;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Factorial;
+
+    procedure Show_Factorial is
+    begin
+       for I in 1 .. 50 loop
+          Put_Line (I'Image & "! = "
+                    & Factorial (I)'Image);
+       end loop;
+    end Show_Factorial;
+
+Here, we're using :ada:`Long_Long_Integer` for the computation and return type
```

Comment:
You might want to say, parenthetically, that the range of Long_Long_Integer is probably the biggest possible on the machine, although that is not necessarily so.

Reply:
You're right; I've just added it.